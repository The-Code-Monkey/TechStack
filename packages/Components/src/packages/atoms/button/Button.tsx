import React, { useMemo } from 'react';

import { Icon, InteractableProps, Text } from '../../primatives';

import { StyledInteractable, iconOrientations, iconMargins } from './styled';

export interface Props extends InteractableProps {
  iconName?: string;
  iconPosition?: 'left' | 'top' | 'right' | 'bottom';
  variant?: string;
}

const Button = ({
  className,
  children = '',
  iconName,
  iconPosition = 'left',
  autoid,
  variant = 'default',
  ...rest
}: Props) => {
  const hasChildren = useMemo(() => {
    if (typeof children === 'string' && children === '') {
      return false;
    }
    return true;
  }, [children]);

  return (
    <StyledInteractable
      className={className}
      forwardedAs="button"
      autoid={`${autoid || children}_button`}
      typography="button"
      variant={variant}
      {...rest}
      {...iconOrientations[iconPosition]}
    >
      {iconName && (
        <Icon
          name={iconName}
          {...(hasChildren &&
            iconPosition in iconMargins &&
            iconMargins[iconPosition])}
        />
      )}
      {typeof children === 'string' ? <Text>{children}</Text> : { children }}
    </StyledInteractable>
  );
};

export default Button;
