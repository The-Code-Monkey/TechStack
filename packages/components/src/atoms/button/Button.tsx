import * as React from 'react';

const { useMemo } = React;

import { Icon, InteractableProps, Text, IconProps } from '../../primal';

import { StyledInteractable, iconOrientations, iconMargins } from './styled';

export interface Props extends InteractableProps {
  iconName?: IconProps['name'];
  iconPosition?: 'left' | 'top' | 'right' | 'bottom';
  variant?: string;
  strong?: boolean;
}

const Button = ({
  className,
  children = '',
  iconName,
  iconPosition = 'left',
  autoid,
  variant = 'default',
  ...rest
}: Props) => {
  const hasChildren = useMemo(
    () => !(typeof children === 'string' && children === ''),
    [children]
  );

  return (
    <StyledInteractable
      className={className}
      forwardedAs='button'
      autoid={`${autoid || children}_button`}
      typography='button'
      variant={variant}
      {...rest}
      {...iconOrientations[iconPosition]}
    >
      {iconName && (
        <Icon
          name={iconName}
          {...(hasChildren &&
            iconPosition in iconMargins &&
            iconMargins[iconPosition])}
        />
      )}
      {typeof children === 'string' ? <Text>{children}</Text> : { children }}
    </StyledInteractable>
  );
};

export default Button;
