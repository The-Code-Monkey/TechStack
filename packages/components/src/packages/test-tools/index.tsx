import { mount, shallow, ShallowWrapper, ReactWrapper } from 'enzyme';
import React, { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components';

import theme from '../theme';
import { ConfigContext } from '../utils';

const config = {
  outputDir: '/src/packages/',
  iconDir: 'feather',
};

const removeProperties = () => {
  const keys = ['theme', 'styledTheme'];
  return {
    test: (val: any) =>
      val &&
      typeof val === 'object' &&
      'props' in val &&
      Object.keys(val.props).some((prop) => keys.some((key) => key === prop)),
    print: (val: any, serialize: any) => {
      keys.forEach((key) => {
        delete val.props[key];
      });
      return serialize(val);
    },
  };
};

const formattedTheme = {
  ...theme,
  colors: {
    common: theme.colors?.common,
    modes: theme.colors?.modes,
    ...theme.colors?.modes.light,
  },
};

export const shallowWithTheme: (children: ReactNode) => ShallowWrapper = (
  children: ReactNode
) =>
  shallow(<ThemeProvider theme={formattedTheme}>{children}</ThemeProvider>)
    .dive()
    .shallow();

export const mountWithTheme: (children: ReactNode) => ReactWrapper = (
  children: ReactNode
) => {
  expect.addSnapshotSerializer(removeProperties());

  return mount(
    <ConfigContext.Provider value={config}>
      <ThemeProvider theme={formattedTheme}>
        <>{children}</>
      </ThemeProvider>
    </ConfigContext.Provider>
  );
};
