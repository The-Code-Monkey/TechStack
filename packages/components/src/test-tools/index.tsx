import isPropValid from '@emotion/is-prop-valid';
import { render, RenderResult } from '@testing-library/react';
import { ReactNode } from 'react';
import { StyleSheetManager, ThemeProvider } from 'styled-components';
import { expect } from 'bun:test';

import theme from '../theme';
import { ConfigContext } from '../utils';

const config = {
  outputDir: '/src/packages/',
  iconDir: 'feather',
};

const removeProperties = () => {
  const keys = ['theme', 'styledTheme'];
  return {
    test: (val: { props: unknown }) =>
      val &&
      typeof val === 'object' &&
      'props' in val &&
      Object.keys(val.props).some(prop => keys.some(key => key === prop)),
    serialize: (val, config, indentation, depth, refs, printer) => {
      keys.forEach(key => {
        delete val.props[key];
      });
      return printer(val, config, indentation, depth, refs);
    },
  };
};

const formattedTheme = {
  ...theme,
  colors: {
    common: theme.colors?.common,
    modes: theme.colors?.modes,
    ...theme.colors?.modes.light,
  },
};

export const mountWithTheme = (children: ReactNode): RenderResult => {
  // expect.addSnapshotSerializer(removeProperties());

  return render(
    <StyleSheetManager shouldForwardProp={isPropValid}>
      <ConfigContext.Provider value={config}>
        <ThemeProvider theme={formattedTheme}>
          <>{children}</>
        </ThemeProvider>
      </ConfigContext.Provider>
    </StyleSheetManager>
  );
};
