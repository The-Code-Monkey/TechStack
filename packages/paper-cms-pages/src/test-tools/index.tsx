import { ConfigContext } from '@techstack/components';
import { render } from '@testing-library/react';
import { ReactNode } from 'react';

import theme from '../theme';
import { ThemeProvider } from '../workarounds/styled-components';

const config = {
  outputDir: '/src/packages/',
  iconDir: 'feather',
};

const removeProperties = () => {
  const keys = ['theme', 'styledTheme'];
  return {
    test: (val: { props: unknown }) =>
      val &&
      typeof val === 'object' &&
      'props' in val &&
      Object.keys(val.props).some(prop => keys.some(key => key === prop)),
    serialize: (val, config, indentation, depth, refs, printer) => {
      keys.forEach(key => {
        delete val.props[key];
      });
      return printer(val, config, indentation, depth, refs);
    },
  };
};

const formattedTheme = {
  ...theme,
  colors: {
    common: theme.colors?.common,
    modes: theme.colors?.modes,
    ...theme.colors?.modes.light,
  },
};

export const mountWithTheme = (children: ReactNode) => {
  expect.addSnapshotSerializer(removeProperties());

  return render(
    <ConfigContext.Provider value={config}>
      <ThemeProvider theme={formattedTheme}>
        <>{children}</>
      </ThemeProvider>
    </ConfigContext.Provider>
  );
};
