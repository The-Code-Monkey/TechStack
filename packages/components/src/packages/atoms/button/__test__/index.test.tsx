import toJson from 'enzyme-to-json';
import React from 'react';

import Button from '..';
import { mountWithTheme } from '../../../test-tools';

describe('<Button />', () => {
  it('renders default correctly', () => {
    const wrapper = mountWithTheme(<Button>Default</Button>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders primary correctly', () => {
    const wrapper = mountWithTheme(<Button variant="primary">Primary</Button>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders secondary correctly', () => {
    const wrapper = mountWithTheme(
      <Button variant="secondary">Secondary</Button>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders icon correctly', () => {
    const wrapper = mountWithTheme(
      <Button iconName="GitHub">Secondary</Button>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders icon only correctly', () => {
    const wrapper = mountWithTheme(<Button iconName="GitHub" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
