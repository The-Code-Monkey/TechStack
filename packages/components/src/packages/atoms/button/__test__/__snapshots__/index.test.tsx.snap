// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders default correctly 1`] = `
<Component>
  <Button>
    <Styled(Interactable)
      autoid="Default_button"
      flexDirection="row"
      forwardedAs="button"
      typography="button"
      variant="default"
    >
      <Interactable
        as="button"
        autoid="Default_button"
        className="sc-jSMfEi hxGaF"
        flexDirection="row"
        typography="button"
        variant="default"
      >
        <Styled(Box)
          as="button"
          className="sc-jSMfEi hxGaF"
          data-autoid="Default_button"
          flexDirection="row"
          onClick={[Function]}
          typography="button"
          variant="default"
        >
          <button
            className="sc-hKMtZM sc-jSMfEi hxGaF"
            data-autoid="Default_button"
            onClick={[Function]}
          >
            <Text>
              <styled.span
                data-autoid=""
              >
                <span
                  className="sc-eCYdqJ fZcZz"
                  data-autoid=""
                >
                  Default
                </span>
              </styled.span>
            </Text>
          </button>
        </Styled(Box)>
      </Interactable>
    </Styled(Interactable)>
  </Button>
</Component>
`;

exports[`<Button /> renders icon correctly 1`] = `
<Component>
  <Button
    iconName="GitHub"
  >
    <Styled(Interactable)
      autoid="Secondary_button"
      flexDirection="row"
      forwardedAs="button"
      typography="button"
      variant="default"
    >
      <Interactable
        as="button"
        autoid="Secondary_button"
        className="sc-jSMfEi hxGaF"
        flexDirection="row"
        typography="button"
        variant="default"
      >
        <Styled(Box)
          as="button"
          className="sc-jSMfEi hxGaF"
          data-autoid="Secondary_button"
          flexDirection="row"
          onClick={[Function]}
          typography="button"
          variant="default"
        >
          <button
            className="sc-hKMtZM sc-jSMfEi hxGaF"
            data-autoid="Secondary_button"
            onClick={[Function]}
          >
            <Memo(Icon)
              ml={-2}
              mr={2}
              name="GitHub"
            >
              <Suspense
                fallback={<ForwardRef(HelpCircle) />}
              >
                <Styled(Box)
                  autoid="github_icon"
                  ml={-2}
                  mr={2}
                >
                  <Box
                    autoid="github_icon"
                    className="sc-dkzDqf huekpF"
                  >
                    <styled.div
                      className="sc-dkzDqf huekpF"
                      data-autoid="github_icon"
                    >
                      <div
                        className="sc-gsnTZi kJlumN sc-dkzDqf huekpF"
                        data-autoid="github_icon"
                      />
                    </styled.div>
                  </Box>
                </Styled(Box)>
                <ForwardRef(HelpCircle)>
                  <svg
                    fill="none"
                    height={24}
                    stroke="currentColor"
                    strokeLinecap="round"
                    strokeLinejoin="round"
                    strokeWidth="2"
                    viewBox="0 0 24 24"
                    width={24}
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <circle
                      cx="12"
                      cy="12"
                      r="10"
                    />
                    <path
                      d="M9.09 9a3 3 0 0 1 5.83 1c0 2-3 3-3 3"
                    />
                    <line
                      x1="12"
                      x2="12.01"
                      y1="17"
                      y2="17"
                    />
                  </svg>
                </ForwardRef(HelpCircle)>
              </Suspense>
            </Memo(Icon)>
            <Text>
              <styled.span
                data-autoid=""
              >
                <span
                  className="sc-eCYdqJ fZcZz"
                  data-autoid=""
                >
                  Secondary
                </span>
              </styled.span>
            </Text>
          </button>
        </Styled(Box)>
      </Interactable>
    </Styled(Interactable)>
  </Button>
</Component>
`;

exports[`<Button /> renders icon only correctly 1`] = `
<Component>
  <Button
    iconName="GitHub"
  >
    <Styled(Interactable)
      autoid="_button"
      flexDirection="row"
      forwardedAs="button"
      typography="button"
      variant="default"
    >
      <Interactable
        as="button"
        autoid="_button"
        className="sc-jSMfEi hxGaF"
        flexDirection="row"
        typography="button"
        variant="default"
      >
        <Styled(Box)
          as="button"
          className="sc-jSMfEi hxGaF"
          data-autoid="_button"
          flexDirection="row"
          onClick={[Function]}
          typography="button"
          variant="default"
        >
          <button
            className="sc-hKMtZM sc-jSMfEi hxGaF"
            data-autoid="_button"
            onClick={[Function]}
          >
            <Memo(Icon)
              name="GitHub"
            >
              <Suspense
                fallback={<ForwardRef(HelpCircle) />}
              >
                <Styled(Box)
                  autoid="github_icon"
                >
                  <Box
                    autoid="github_icon"
                    className="sc-dkzDqf bQHrml"
                  >
                    <styled.div
                      className="sc-dkzDqf bQHrml"
                      data-autoid="github_icon"
                    >
                      <div
                        className="sc-gsnTZi kJlumN sc-dkzDqf bQHrml"
                        data-autoid="github_icon"
                      />
                    </styled.div>
                  </Box>
                </Styled(Box)>
                <ForwardRef(HelpCircle)>
                  <svg
                    fill="none"
                    height={24}
                    stroke="currentColor"
                    strokeLinecap="round"
                    strokeLinejoin="round"
                    strokeWidth="2"
                    viewBox="0 0 24 24"
                    width={24}
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <circle
                      cx="12"
                      cy="12"
                      r="10"
                    />
                    <path
                      d="M9.09 9a3 3 0 0 1 5.83 1c0 2-3 3-3 3"
                    />
                    <line
                      x1="12"
                      x2="12.01"
                      y1="17"
                      y2="17"
                    />
                  </svg>
                </ForwardRef(HelpCircle)>
              </Suspense>
            </Memo(Icon)>
            <Text>
              <styled.span
                data-autoid=""
              >
                <span
                  className="sc-eCYdqJ fZcZz"
                  data-autoid=""
                />
              </styled.span>
            </Text>
          </button>
        </Styled(Box)>
      </Interactable>
    </Styled(Interactable)>
  </Button>
</Component>
`;

exports[`<Button /> renders primary correctly 1`] = `
<Component>
  <Button
    variant="primary"
  >
    <Styled(Interactable)
      autoid="Primary_button"
      flexDirection="row"
      forwardedAs="button"
      typography="button"
      variant="primary"
    >
      <Interactable
        as="button"
        autoid="Primary_button"
        className="sc-jSMfEi fOiwAw"
        flexDirection="row"
        typography="button"
        variant="primary"
      >
        <Styled(Box)
          as="button"
          className="sc-jSMfEi fOiwAw"
          data-autoid="Primary_button"
          flexDirection="row"
          onClick={[Function]}
          typography="button"
          variant="primary"
        >
          <button
            className="sc-hKMtZM sc-jSMfEi fOiwAw"
            data-autoid="Primary_button"
            onClick={[Function]}
          >
            <Text>
              <styled.span
                data-autoid=""
              >
                <span
                  className="sc-eCYdqJ fZcZz"
                  data-autoid=""
                >
                  Primary
                </span>
              </styled.span>
            </Text>
          </button>
        </Styled(Box)>
      </Interactable>
    </Styled(Interactable)>
  </Button>
</Component>
`;

exports[`<Button /> renders secondary correctly 1`] = `
<Component>
  <Button
    variant="secondary"
  >
    <Styled(Interactable)
      autoid="Secondary_button"
      flexDirection="row"
      forwardedAs="button"
      typography="button"
      variant="secondary"
    >
      <Interactable
        as="button"
        autoid="Secondary_button"
        className="sc-jSMfEi bWHMQb"
        flexDirection="row"
        typography="button"
        variant="secondary"
      >
        <Styled(Box)
          as="button"
          className="sc-jSMfEi bWHMQb"
          data-autoid="Secondary_button"
          flexDirection="row"
          onClick={[Function]}
          typography="button"
          variant="secondary"
        >
          <button
            className="sc-hKMtZM sc-jSMfEi bWHMQb"
            data-autoid="Secondary_button"
            onClick={[Function]}
          >
            <Text>
              <styled.span
                data-autoid=""
              >
                <span
                  className="sc-eCYdqJ fZcZz"
                  data-autoid=""
                >
                  Secondary
                </span>
              </styled.span>
            </Text>
          </button>
        </Styled(Box)>
      </Interactable>
    </Styled(Interactable)>
  </Button>
</Component>
`;
